max-chelsea/starter-code/public/scripts/article.js

```diff
@@ -0,0 +1,159 @@
+      return firstNum + secondNum;
+    })
+  };
+
+  // Done: Chain together a `map` and a `reduce` call to produce an array of unique author names. You will
+  // probably need to use the optional accumulator argument in your reduce call.
+  Article.allAuthors = () => {
+    return Article.all
+    .map(function(article){
+      return article.author;
+    })
+    .reduce(function(aggregator, authors){
+      if(!aggregator.includes(authors)){
+        return aggregator.push(authors);
+      }
+    }, [])
+  };
+
+  Article.numWordsByAuthor = () => {
+    return Article.allAuthors().map(author => {
+      return {'author': author, 'wordCount' : Article.all
+      .filter(function(articles){
+        return articles.author = author;
+      })
+      .reduce(function(total, currentValue) {
+        return currentValue.body.split(' ').length + total;
+      }, 0)};
+
+      // Done: Transform each author string into an object with properties for
+      // the author's name, as well as the total number of words across all articles
+      // written by the specified author.
+      // HINT: This .map should be setup to return an object literal with two properties.
+      // The first property should be pretty straightforward, but you will need to chain
+      // some combination of filter, map, and reduce to get the value for the second
+      // property.
+      // gives us an array of all unique author names
+
+    })
+  };
+
+  Article.truncateTable = callback => {
+    $.ajax({
+      url: '/articles',
+      method: 'DELETE',
+    })
+    .then(console.log) // REVIEW: Check out this clean syntax for just passing 'assumed' data into a named function!
+                       // The reason we can do this has to do with the way Promise.prototype.then works. It's a little
+                       // outside the scope of 301 material, but feel free to research!
+    .then(callback);
+  };
+
+  Article.prototype.insertRecord = function(callback) {
+    // REVIEW: Why can't we use an arrow function here for .insertRecord()??
+    $.post('/articles', {author: this.author, authorUrl: this.authorUrl, body: this.body, category: this.category, publishedOn: this.publishedOn, title: this.title})
+    .then(console.log)
+    .then(callback);
+  };
+
+  Article.prototype.deleteRecord = function(callback) {
+    $.ajax({
+      url: `/articles/${this.article_id}`,
+      method: 'DELETE'
+    })
+    .then(console.log)
+    .then(callback);
+  };
+
+  Article.prototype.updateRecord = function(callback) {
+    $.ajax({
+      url: `/articles/${this.article_id}`,
+      method: 'PUT',
+      data: {
+        author: this.author,
+        authorUrl: this.authorUrl,
+        body: this.body,
+        category: this.category,
+        publishedOn: this.publishedOn,
+        title: this.title,
+        author_id: this.author_id
+      }
+    })
+    .then(console.log)
+    .then(callback);
+  };
+  module.Article = Article;
+})();
```

Comment:
Remember it's a function of app. We must pass in app which we refer to as module inside the function. (app is an object where your functions/variables will live since we said module.Article = Article and module outside of the function is called app, that's why we can access Article.all via app.Article.all in the console)